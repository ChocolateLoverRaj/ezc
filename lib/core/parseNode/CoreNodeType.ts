enum CoreNodeType {
  FILE,
  DECLARE,
  FUNCTION,
  INTEGER_TYPE,
  GLOBAL_VARIABLE,
  STRING,
  FLOAT_TYPE,
  POINTER_TYPE,
  ARRAY_TYPE,
  STRUCT_TYPE,
  IDENTIFIER,
  NUMBER,
  VOID_TYPE,
  BLOCK,
  RETURN_INSTRUCTION,
  CALL_ASSIGNABLE,
  ASSIGNMENT_INSTRUCTION
}

export default CoreNodeType
