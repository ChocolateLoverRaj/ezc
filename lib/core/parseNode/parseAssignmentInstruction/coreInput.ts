import coreParseCallAssignableInput from '../parseCallAssignable/coreInput'
import parseCallAssignable from '../parseCallAssignable/parseCallAssignable'
import Input from './Input'

const coreInput: Input = [parseCallAssignable(coreParseCallAssignableInput)]

export default coreInput
