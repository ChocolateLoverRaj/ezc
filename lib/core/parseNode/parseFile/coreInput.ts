import coreParseDeclareInput from '../parseDeclare/coreInput'
import parseDeclare from '../parseDeclare/parseDeclare'
import parseFunction from '../parseFunction/parseFunction'
import coreParseGlobalVariableInput from '../parseGlobalVariable/coreInput'
import parseGlobalVariable from '../parseGlobalVariable/parseGlobalVariable'
import tryNodeParsers from '../tryNodeParsers'
import Input from './Input'
import coreParseFunctionInput from '../parseFunction/coreInput'

const coreInput: Input = tryNodeParsers([
  parseDeclare(coreParseDeclareInput),
  parseGlobalVariable(coreParseGlobalVariableInput),
  parseFunction(coreParseFunctionInput)
])

export default coreInput
