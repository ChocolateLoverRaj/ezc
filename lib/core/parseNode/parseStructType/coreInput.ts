import coreTypeParsers from '../coreTypeParsers'
import Input from './Input'

const coreInput: Input = coreTypeParsers

export default coreInput
