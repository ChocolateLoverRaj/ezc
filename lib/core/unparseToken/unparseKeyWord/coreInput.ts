import CoreKeyWord from '../../CoreKeyWord'
import Input from './Input'

const coreKeyWordToStringMap: Map<CoreKeyWord, string> = new Map([
  [CoreKeyWord.ALIGN, 'align'],
  [CoreKeyWord.CALL, 'call'],
  [CoreKeyWord.COMMA, ','],
  [CoreKeyWord.CONSTANT, 'constant'],
  [CoreKeyWord.DECLARE, 'declare'],
  [CoreKeyWord.DEFINE, 'define'],
  [CoreKeyWord.EQUALS, '='],
  [CoreKeyWord.GET_ELEMENT_PTR, 'getelementptr'],
  [CoreKeyWord.GLOBAL, 'global'],
  [CoreKeyWord.INBOUNDS, 'inbounds'],
  [CoreKeyWord.NO_ALIAS, 'noalias'],
  [CoreKeyWord.NO_CAPTURE, 'nocapture'],
  [CoreKeyWord.NO_FREE, 'nofree'],
  [CoreKeyWord.PRIVATE, 'private'],
  [CoreKeyWord.PTR, 'ptr'],
  [CoreKeyWord.RETURN, 'ret'],
  [CoreKeyWord.UNNAMED_ADDR, 'unnamed_addr'],
  [CoreKeyWord.VOID, 'void'],
  [CoreKeyWord.X, 'x'],
  [CoreKeyWord.OPEN_PARENTHESIS, '('],
  [CoreKeyWord.CLOSE_PARENTHESIS, ')'],
  [CoreKeyWord.OPEN_BRACKET, '['],
  [CoreKeyWord.CLOSE_BRACKET, ']'],
  [CoreKeyWord.OPEN_CURLY_BRACKET, '{'],
  [CoreKeyWord.CLOSE_CURLY_BRACKET, '}']
])

const coreInput: Input = new Map([
  [CoreKeyWord, coreKeyWordToStringMap]
])

export default coreInput
