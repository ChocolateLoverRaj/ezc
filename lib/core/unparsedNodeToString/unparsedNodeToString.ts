import UnparsedPartType from '../unparseNode/UnparsedPartType'
import Input from './Input'
import unparseToken from '../unparseToken/unparseToken/unparseToken'
import UnparsedPart from '../unparseNode/UnparsedPart'

const indentSpaces = 2
const indentStr = ' '.repeat(indentSpaces)
const unparsedNodeToString = ({ nodeUnparsers, unparseTokenInput }: Input) => (
  unparsedParts: readonly UnparsedPart[]
): string => {
  let indentLevel = 0
  let string = ''
  const partsToUnparse: UnparsedPart[] = [...unparsedParts]
  while (partsToUnparse.length > 0) {
    const { type, data } = partsToUnparse[0]
    partsToUnparse.shift()
    if (type === UnparsedPartType.NEW_LINE) {
      string += `\n${indentStr.repeat(indentLevel)}`
    } else if (type === UnparsedPartType.INDENT) {
      string += indentStr
      indentLevel++
    } else if (type === UnparsedPartType.UNINDENT) {
      indentLevel--
    } else if (type === UnparsedPartType.SPACE) {
      string += ' '
    } else if (type === UnparsedPartType.TOKEN) {
      string += unparseToken(unparseTokenInput)(data)
    } else if (type === UnparsedPartType.NODE) {
      const unparsedParts =
        nodeUnparsers.get(data.type.enum)?.get(data.type.id)?.(data.data) as UnparsedPart[]
      if (unparsedParts === undefined) console.log(data)
      partsToUnparse.unshift(...unparsedParts)
    }
  }

  return string
}

export default unparsedNodeToString
