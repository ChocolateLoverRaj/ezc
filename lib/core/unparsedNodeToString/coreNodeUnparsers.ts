import CoreNodeType from '../parseNode/CoreNodeType'
import unparseArrayType from '../unparseNode/unparseArrayType'
import unparseDeclare from '../unparseNode/unparseDeclare/unparseDeclare'
import unparseFile from '../unparseNode/unparseFile'
import coreUnparseInputFlagInput from '../unparseNode/unparseInputFlag/coreInput'
import unparseInputFlag from '../unparseNode/unparseInputFlag/unparseInputFlag'
import unparseIntegerType from '../unparseNode/unparseIntegerType'
import UnparseNode from '../unparseNode/UnparseNode'
import unparsePointerType from '../unparseNode/unparsePointerType'
import unparseString from '../unparseNode/unparseString'
import unparseVoidType from '../unparseNode/unparseVoidType'
import NodeUnparsers from './NodeUnparsers'

const coreNodeUnparsers: NodeUnparsers = new Map([
  [CoreNodeType, new Map<number, UnparseNode<unknown>>([
    [CoreNodeType.DECLARE, unparseDeclare],
    [CoreNodeType.INPUT_FLAG, unparseInputFlag(coreUnparseInputFlagInput)],
    [CoreNodeType.ARRAY_TYPE, unparseArrayType],
    [CoreNodeType.FILE, unparseFile],
    [CoreNodeType.STRING, unparseString],
    [CoreNodeType.VOID_TYPE, unparseVoidType],
    [CoreNodeType.INTEGER_TYPE, unparseIntegerType],
    [CoreNodeType.POINTER_TYPE, unparsePointerType]
  ])]
])

export default coreNodeUnparsers
