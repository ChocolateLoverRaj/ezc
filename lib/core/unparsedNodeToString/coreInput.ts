import coreUnparseTokenInput from '../unparseToken/unparseToken/coreInput'
import coreNodeUnparsers from './coreNodeUnparsers'
import Input from './Input'

const coreInput: Input = {
  unparseTokenInput: coreUnparseTokenInput,
  nodeUnparsers: coreNodeUnparsers
}

export default coreInput
