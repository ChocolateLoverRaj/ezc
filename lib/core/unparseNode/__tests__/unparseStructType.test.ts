import CoreNodeType from '../../parseNode/CoreNodeType'
import coreInput from '../../unparsedNodeToString/coreInput'
import unparsedNodeToString from '../../unparsedNodeToString/unparsedNodeToString'
import unparseStructType from '../unparseStructType'

test('{ i64, ptr }', () => {
  expect(unparsedNodeToString(coreInput)(unparseStructType([{
    type: {
      enum: CoreNodeType,
      id: CoreNodeType.INTEGER_TYPE
    },
    data: 64
  }, {
    type: {
      enum: CoreNodeType,
      id: CoreNodeType.POINTER_TYPE
    },
    data: undefined
  }]))).toMatchSnapshot()
})

test('{ i32 }', () => {
  expect(unparsedNodeToString(coreInput)(unparseStructType([{
    type: {
      enum: CoreNodeType,
      id: CoreNodeType.INTEGER_TYPE
    },
    data: 32
  }]))).toMatchSnapshot()
})

test('{ i1, i64, ptr }', () => {
  expect(unparsedNodeToString(coreInput)(unparseStructType([{
    type: {
      enum: CoreNodeType,
      id: CoreNodeType.INTEGER_TYPE
    },
    data: 1
  }, {
    type: {
      enum: CoreNodeType,
      id: CoreNodeType.INTEGER_TYPE
    },
    data: 64
  }, {
    type: {
      enum: CoreNodeType,
      id: CoreNodeType.POINTER_TYPE
    },
    data: undefined
  }]))).toMatchSnapshot()
})
