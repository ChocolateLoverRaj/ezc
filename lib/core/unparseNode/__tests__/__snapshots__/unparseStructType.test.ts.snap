// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`{ i1, i64, ptr } 1`] = `
"{  
  i1,
  i64,
  ptr
}"
`;

exports[`{ i32 } 1`] = `
"{  
  i32
}"
`;

exports[`{ i64, ptr } 1`] = `
"{  
  i64,
  ptr
}"
`;
